Add MFA for Windows Desktop web access

smallinsky opened the pull request:
## What: 
Add backend support for Desktop Access MFA. 

## How:
- Reused Node MFA mechanism where MFA challange message  is encoded by protobuf and send to the client over web socket.  
- Support of OPT is missing because node MFA at the moment supports only `U2F` and `webauthn`. 
- during MFA logic errors are propagated to the client by tdp.Error  message.

smallinsky committed:
Add MFA for Windows Desktop web access

smallinsky committed:
Merge remote-tracking branch 'origin/master' into smallinsky/mfa-for-rds-v2

smallinsky committed:
Update proto

smallinsky committed:
update

smallinsky committed:
Update proto comments

smallinsky committed:
update

smallinsky committed:
update

smallinsky committed:
update

zmb3 commented on lib/client/interfaces.go:
```diff
@@ -86,7 +86,9 @@ type Key struct {
 	AppTLSCerts map[string][]byte `json:"AppCerts,omitempty"`
-
+	// WindowsDesktopCerts are TLS certificates for RDS Windows Desktop access.
+	// Map key is the RDS server name.
```
```suggestion
	// Map key is the desktop name.
```

r0mant reviewed (commented):
Don't have comments beyond what Zac's already pointed out.

smallinsky committed:
update

smallinsky committed:
Merge remote-tracking branch 'origin/master' into smallinsky/mfa-for-rds-v2

smallinsky commented on lib/auth/grpcserver.go:
```diff
@@ -2031,6 +2031,7 @@ func validateUserSingleUseCertRequest(ctx context.Context, actx *grpcContext, re
 		}
 	case proto.UserCertsRequest_All:
 		return trace.BadParameter("must specify a concrete Usage in UserCertsRequest, one of SSH, Kubernetes or Database")
+	case proto.UserCertsRequest_WindowsDesktop:
```
Added

smallinsky commented on lib/client/client.go:
```diff
@@ -485,6 +496,23 @@ func (proxy *ProxyClient) prepareUserCertsRequest(params ReissueParams, key *Key
 func (proxy *ProxyClient) RootClusterName() (string, error) {
 	tlsKey, err := proxy.localAgent().GetCoreKey()
 	if err != nil {
+		if trace.IsNotFound(err) {
```
yes, I have found this issue when I tried to call pc.RootClusterName() in Windows Access flow when the proxy client don't contain local agent. In current implemeantion this doesn't happen I can remove it and rise a seperate PR with this improvement.

zmb3 replied:
I don't have a problem with including here, just wanted to double check what it was for.

smallinsky commented on lib/web/desktop.go:
```diff
@@ -117,30 +113,110 @@ func createDesktopConnection(
+		return nil, trace.Wrap(err)
+	}
+
+	key, err := pc.IssueUserCertsWithMFA(ctx, client.ReissueParams{
```
The `IssueUserCertsWithMFA` distinguish if MFA is enabled based on   `clt.IsMFARequired` grpc result returned by auth server. If MFA is disabled to "tap" logic is skipped.

smallinsky committed:
update

smallinsky commented on go.mod:
```diff
@@ -38,6 +38,7 @@ require (
 	github.com/gokyle/hotp v0.0.0-20160218004637-c180d57d286b
 	github.com/golang/protobuf v1.5.2
 	github.com/google/btree v1.0.0
+	github.com/google/certificate-transparency-go v1.0.21
```
Not sure what happened here but you are right the `certificate-transparency-go` is indirect : 
```sh
$ go mod why github.com/google/certificate-transparency-go
# github.com/google/certificate-transparency-go
github.com/gravitational/teleport/lib/auth/mocku2f
github.com/duo-labs/webauthn/protocol
github.com/duo-labs/webauthn/metadata
github.com/cloudflare/cfssl/revoke
github.com/cloudflare/cfssl/helpers
github.com/google/certificate-transparency-go
```

Fixed.

ibeckermayer commented on lib/web/desktop.go:
```diff
@@ -117,30 +113,110 @@ func createDesktopConnection(
 }
 
+func writeError(err error, ws *websocket.Conn, log *logrus.Entry) {
+	rdpConn := tdp.NewConn(ws)
+	oerr := rdpConn.OutputMessage(tdp.Error{Message: err.Error()})
```
```suggestion
	tdpConn := tdp.NewConn(ws)
	oerr := tdpConn.OutputMessage(tdp.Error{Message: err.Error()})
```

zmb3 replied:
`*tdp.Conn` now has a `SendError(string)` method, so you can just do:

```
tdp.NewConn(ws).SendError(err.Error())
```

ibeckermayer reviewed (approved):
Just a small inline suggestion but otherwise lgtm

smallinsky committed:
Merge remote-tracking branch 'origin/master' into smallinsky/mfa-for-rds-v2

smallinsky committed:
update

ibeckermayer commented on lib/srv/desktop/windows_server.go:
```diff
@@ -788,9 +786,13 @@ func (s *WindowsService) connectRDP(ctx context.Context, log logrus.FieldLogger,
 		windowsUser = login // capture attempted login user
+		mfaParams := services.AccessMFAParams{
+			Verified:       identity.MFAVerified != "",
+			AlwaysRequired: authPref.GetRequireSessionMFA(),
```
The only implementation for this my editor finds is [here](https://github.com/gravitational/teleport/blob/a943a5f09dcd0417eff52221d0485013a818ca13/api/types/authentication.go#L345-L349), which returns the field corresponding to this definition [here](https://github.com/gravitational/teleport/blob/a943a5f09dcd0417eff52221d0485013a818ca13/api/types/types.proto#L993-L995), which (if I'm not mistaken) corresponds to the [cluster-wide per-session MFA setting](https://goteleport.com/docs/access-controls/guides/per-session-mfa/#cluster-wide). However according to that documentation, we also support [per-role mfa](https://goteleport.com/docs/access-controls/guides/per-session-mfa/#per-role) which this doesn't seem to account for.

zmb3 replied:
I think this is fine. `AlwaysRequired` means is enabled at the cluster level, so there's no need to check the roles.

If you look at `CheckAccess()` in lib/services/role.go, you'll see we also consult the roles if `AlwaysRequired` was false:

```
// if MFA is not verified and we require session MFA, deny access
if role.GetOptions().RequireSessionMFA {
```

smallinsky committed:
Merge remote-tracking branch 'origin/master' into smallinsky/mfa-for-rds-v2

smallinsky committed:
update

Commit:
Add MFA message to TDP (#10439)

* adds an MfaJson message type to TDP and refactors promptMFAChallenge to be able to accept functions for encode and decode so that it can be used for both protobuf and tdp/json messages

* Add TDP message 10 (MFA)

Per-session MFA works just like it does for SSH sessions, but rather
than wrapping the MFA JSON challenge in a protobuf (like SSH), we
package it in a TDP message.

* Create MFA codec interface

* Fix build

* Reduce some JSON decoding duplication

* General cleanup

Co-authored-by: Zac Bergquist <zac.bergquist@goteleport.com>

smallinsky committed:
Merge remote-tracking branch 'origin/master' into smallinsky/mfa-for-rds-v2

smallinsky committed:
Update

zmb3 commented on lib/srv/desktop/tdp/proto.go:
```diff
@@ -443,6 +453,155 @@ func decodeClipboardData(in peekReader, maxLen uint32) (ClipboardData, error) {
+		return nil, trace.Wrap(err)
+	}
+	if t != byte(TypeMFA) {
+		return nil, trace.BadParameter("got message type %v, expected TypeMFAJson(%v)", t, TypeMFA)
```
```suggestion
		return nil, trace.BadParameter("got message type %v, expected TypeMFA(%v)", t, TypeMFA)
```

my rename didn't catch this one

zmb3 commented on lib/srv/desktop/tdp/proto.go:
```diff
@@ -443,6 +453,155 @@ func decodeClipboardData(in peekReader, maxLen uint32) (ClipboardData, error) {
+	}, nil
+}
+
+func DecodeMFAChalange(in peekReader) (*MFA, error) {
```
Is this for debugging or something?

smallinsky replied:
For unit test, we need to have ability to Decode changed from go code. Right noe MFA Decode allows to decode MFA Response type because the same message is used for MFAChallnage and MFAChallnage Response and right now codec don't provide ability to decode `MFAChallange`

smallinsky committed:
update

smallinsky committed:
udpate

smallinsky committed:
Update

smallinsky committed:
update

smallinsky committed:
Merge branch 'master' into smallinsky/mfa-for-rds-v2

r0mant commented:
@ibeckermayer Could you take another look and approve if your feedback is sufficiently addressed? Thanks.

smallinsky committed:
Add missing ws error propagation

smallinsky committed:
fix

smallinsky committed:
fix wrong errors check

ibeckermayer commented on lib/srv/desktop/tdp/proto.go:
```diff
@@ -443,6 +454,159 @@ func decodeClipboardData(in peekReader, maxLen uint32) (ClipboardData, error) {
+	}, nil
+}
+
+func DecodeMFAChalange(in peekReader) (*MFA, error) {
```
```suggestion
func DecodeMFAChallenge(in peekReader) (*MFA, error) {
```

ibeckermayer replied:
Could also use a note that this is for testing purposes only

smallinsky committed:
Merge branch 'master' into smallinsky/mfa-for-rds-v2

smallinsky committed:
rename update comment

smallinsky committed:
Merge branch 'master' into smallinsky/mfa-for-rds-v2

smallinsky committed:
Merge branch 'master' into smallinsky/mfa-for-rds-v2
